src/satisfia/agents/makeMDPAgentSatisfia.py

```diff
@@ -17,1546 +17,1522 @@
+                                   self.world.transition_probability(state, otherActionAndAleph[0], nextState)[0]))
+        def X(actionAndAleph):
+            action, aleph4action = actionAndAleph
+            return self.world.expectation(state, action, Y, (action,)) + self.causation_action(state, aleph4state, action, aleph4action) # recursion
+        res = locPol.expectation(X)
+        if self.debug or self.verbose:
+            print(pad(state),"╰ causation_state", prettyState(state), aleph4state, ":", res)
+        return res
+
+    @lru_cache(maxsize=None)
+    def causationPotential_state(self, state, aleph4state): # recursive
+        """Maximal directed information from action sequence to state sequence over all possible policies"""
+        raise NotImplementedError("causationPotential_state is not yet implemented correctly")
+        if self.debug:
+            print(pad(state),"causationPotential_state", prettyState(state), aleph4state, "...")
+        locPol = self.localPolicy(state, aleph4state)
+        def Y(nextState, action):
+            p = self.world.transition_probability(state, action, nextState)[0]
+            if p == 0: return float("-inf")
+            return math.log(p / locPol.expectation(lambda otherActionAndAleph: 
+                                   self.world.transition_probability(state, otherActionAndAleph[0], nextState)[0]))
+        res = max([self.world.expectation(state, action, Y, (action,)) 
+                    + self.causationPotential_action(state, aleph4state, action, self.aspiration4action(state, action, aleph4state))
+                    for action in self.world.possible_actions(state)]) 
+        if self.debug or self.verbose:
+            print(pad(state),"╰ causationPotential_state", prettyState(state), aleph4state, ":", res)
+        return res
+
+    @lru_cache(maxsize=None)
+    def otherLoss_state(self, state, aleph4state): # recursive
+        if self.debug:
+            print(pad(state),"otherLoss_state", prettyState(state), aleph4state, "...")
+        def X(actionAndAleph):
+            return self.otherLoss_action(state, actionAndAleph[0], actionAndAleph[1]) # recursion
+        res = self.localPolicy(state, aleph4state).expectation(X)
+        if self.debug or self.verbose:
+            print(pad(state),"╰ otherLoss_state", prettyState(state), aleph4state, ":", res)
+        return res
+
+    @cache
+    def randomTieBreaker(self, state, action):
+        return random.random()
+
+    # now we can combine all of the above quantities to a combined (safety) loss function:
+
+    # state, action, aleph4state, aleph4action, estActionProbability
+    @lru_cache(maxsize=None)
+    def combinedLoss(self, s, a, al4s, al4a, p): # recursive
+
+        if self.debug:
+            print(pad(s),"| | combinedLoss, state",prettyState(s),"action",a,"aleph4state",al4s,"aleph4action",al4a,"estActionProbability",p,"...")
+
+        res = sum(k * v(s, a, al4s, al4a, p) for k, v in self.losses.items())
```

Comment:
shouldn't we also make sure that `v` is only called if `k != 0`, like so: `0 if k == 0 else k * v(...)`?